server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -202,13 +202,20 @@ public void updateMapping(final IndexMetadata currentIndexMetadata, final IndexM
     }
 
     private boolean assertRefreshIsNotNeeded(DocumentMapper currentMapper,
-                                          String type,
-                                          CompressedXContent incomingMappingSource,
-                                          Mapping incomingMapping) {
+                                             String type,
+                                             Mapping incomingMapping) {
         Mapping mergedMapping = mergeMappings(currentMapper, incomingMapping, MergeReason.MAPPING_RECOVERY);
+
+        ToXContent.MapParams params = new ToXContent.MapParams(Collections.singletonMap("skip_runtime_section", "true"));
         CompressedXContent mergedMappingSource;
         try {
-            mergedMappingSource = new CompressedXContent(mergedMapping, XContentType.JSON, ToXContent.EMPTY_PARAMS);
+            mergedMappingSource = new CompressedXContent(mergedMapping, XContentType.JSON, params);
+        } catch (Exception e) {
+            throw new AssertionError("failed to serialize source for type [" + type + "]", e);
+        }
+        CompressedXContent incomingMappingSource;
+        try {
+            incomingMappingSource = new CompressedXContent(incomingMapping, XContentType.JSON, params);
```

Comment:
There were a couple of ways to work around this. The assertion was recently added (as the refresh mapping action was just removed) hence it is important to keep it around for some time. It seems rather safe to exclude runtime fields from the check. As for how to exclude them, we could have parsed them back into a map and filter its content, or recreate Mapping with a modified root object mapper that has empty runtime fields. Both of these approaches ended up being more complex that I initially considered. 

This approach has the risk of introducing a custom toxcontent param, which must never be used outside of this context, but it does its job and is relatively simple. I am curious to hear other's thoughts on this. @DaveCTurner @ywelsch too  ;)